rfcs/0042-networking-refactor.md

```diff
@@ -0,0 +1,337 @@
+- gossip and response validation
+  - `handle{Gossip,Response}` message is sent to daemon
+  - `validate` is sent back to helper
+- request validation
+  - `handleRequest` message is sent to daemon
+  - `sendResponse` message is sent back to helper, which contains both the response and validation state
+
+### Staging the Compatible and Hard Fork Changes
+
+Some of the key changes proposed in this RFC require a hard fork in order to be released to the network. However, the next hard fork may be a while out. We could just implement this work off of `develop` and wait until the next hard fork to ship it, but this would mean that any immediate improvements we make to the networking code on `compatible` will conflict with our refactor work on `develop`. Overall, it is still a benefit to have this refactor on `compatible` so that we can benefit from it immediately in the soft fork world while keeping the code more or less in line with the future hard fork we want to take.
+
+Accordingly, in order to break this work up, we can do this refactor in 2 passes: first, perform the main module and organization refactor off of `compatible`, then once that is complete and merged, perform the hard fork specific refactor off of `develop`. The `compatible` portion of the refactor can include all changes that do not effect or rely on changes to messages sent over the gossip network. Below is a detailed list of what would be included in each phase of the refactor.
+
+- `compatible`
+  - Transport layer abstraction
+  - OCaml module conslidation w/ new interface
+  - Daemon/Libp2p protocol refactor (peer abstraction et al)
+  - Validation, Response, Gossip, and Request pipe split
+- `develop`
+  - Message type awareness
+  - Message type based pipe split
+  - Trust scoring based peer-selection (requires trust system)
+
+### OCaml Implementation
+
+Structure wise, the OCaml implementation will continue to model the network interface abstractly so that a dummy implementation may be used for unit testing purposes. We will continue to have an [interface](../src/lib/gossip_net/intf.ml) along with a [existential wrapper](../src/lib/gossip_net/any.ml) that provides indirection over the selected gossip network implementation. A [stub implementation](../src/lib/gossip_net/fake.ml) will also continue to exist.
+
+In order to maintain reasonable separation of concerns, the libp2p helper implementation of the gossip network interface will be split into 2 main modules.
+- `Libp2p` :: direct low-level access to `libp2p_helper` process management and protocol
+- `Mina_net` :: high-level networking interface which defines supported RPCs and exposes networking functionality to the rest of the code (publicly exposed to the rest of the code)
+
+The RPC interface would continue to be defined under the [current GADT based setup](../src/lib/mina_networking/mina_networking.ml). This type setup will also be extended so that `Rpc.implementation`
```

Comment:
Will this remove the redundancy in the current `mina_networking.ml` file? It looks like the extra functions that trivially pattern-match are mostly just doing so to deal with RPC.implementations defined elsewhere? I don't 100% understand what you're saying will change here (and I don't think I need to), just want to make sure you're thinking about trying to limit that boilerplate extra stuff

Reply:
Yeah, plan is to collapse all the current files into the structure defined in this RFC.